WiP starting point for a glossary (#31)
BitcoinDesign/Guide: guide/glossary.md

---
layout: guide
title: Glossary
nav_order: 4
---

# Glossary
{: .no_toc }

## Address

> An identifier of 26-35 alphanumeric characters that can be used to receive bitcoin. There are [several address formats](https://en.bitcoin.it/wiki/List_of_address_prefixes){:target="_blank"} based on different standards and use cases. Applications should make it clear to users which standards are being used. Support for formats varies, so users may need to know this information for backup and future recovery.

Reviewer: ```suggestion
An identifier of 26-35 alphanumeric characters that is used to receive bitcoin. There are [several address formats](https://en.bitcoin.it/wiki/List_of_address_prefixes){:target="_blank"} based on different specifications. Applications should inform users which format it uses as support in varies across applications. Users need to know this information during backup for future recovery.
```

NOTE: In some places we say "wallet-products" and here we say "application". So we should have a conversation on what this should be called to have more harmony in the content - as I've also been leaning towards "applications", "bitcoin applications", "experiences" which introduces more inconsistencies.

Reviewer 2: I am changing product --> application where I encounter it

> When users enter addresses, these formats can also be used to instantly validate correctness and possibly deduct other useful information.

Reviewer: ```suggestion
When users enter an address, these formats have specific prefixes so it is possible to instantly determine which format is being used.
```

These are the three most common ones:

**Legacy address - [P2PKH](https://en.bitcoin.it/wiki/Transaction#Pay-to-PubkeyHash){:target="_blank"}**<br/>
Stands for pay-to-pubkey-hash, i.e pay to a hash of the recipient’s public key. This was the original address format for bitcoin and is often called *legacy address*.

> *Legacy addresses* begin with the number 1 and are mixed and case sensitive. <br/>

Reviewer: ```suggestion
*Legacy addresses* begin with the number 1 can contain upper and lowercase characters and is case sensitive. <br/>
```

Reviewer 2: Will commit and fix typo (is/are)

> Stands for pay-to-pubkey-hash, i.e pay to a hash of the recipient’s public key. This was the original address format for bitcoin and is sometimes called *legacy address*. Legacy addresses are not segwit compatible, but you can still send BTC from a P2PKH address to a segwit address without any problems. The average fee when sending from a P2PKH address is likely to be higher than when sending from a segwit address, however, because legacy address transactions are larger in size.

Author: Not sure what `segwit compatible` means at this point. I think I can figure it out from reading the other definitions, but not totally sure and I have to read between the lines.

Can we simplify these last two sentences to just say that sending from Legacy addresses is a bit more expensive than sending from SegWit addresses, because they can't be stored/compressed as efficiently? I'd be interested in how big those savings are. If it's 0.0001%, then maybe need to say that as it should not be a major consideration.

Should we make some type of recommendation that users should migrate wallets to SegWit because legacy addresses may not be supported in the future, or anything if that sort?

Author: Also, I'd consistently use the term `Legacy` address over `P2PKH`, which is much harder to parse.

Reviewer 2: Simplified by cutting the part that was not related to legacy

For example: `1BvBMSEYstWetqTFn5Au4m4GFg7xJaNVN2`

**Script address - [P2SH](https://en.bitcoin.it/wiki/Pay_to_script_hash){:target="_blank"}**<br/>
A script address can have code attached to enable additional rules and functionality. P2SH stands for pay-to-script-hash, and was defined in [BIP13](https://github.com/bitcoin/bips/blob/master/bip-0013.mediawiki).  Script addresses are most commonly used for multi-sig addresses which can specify, for example, that signatures from several keys are required to authorize the transaction.

> *Script addresses* start with the number 3 and are mixed and case sensitive. <br/>

Reviewer: ```suggestion
*Script addresses* start with the number 3, can contain upper and lowercase characters, and is case sensitive. <br/>
```

> P2SH, which stands for pay-to-script-hash, was defined in [BIP13](https://github.com/bitcoin/bips/blob/master/bip-0013.mediawiki) and enables more functionality than legacy addresses. The P2SH script function is most commonly used for multisig addresses which can specify, for example, that multiple digital signatures are required to authorize the transaction. They are also used to enable non-native segwit transactions using a process known as P2WPKH-in-P2SH. This address type is widely supported and can be used to send funds to both P2PKH and Bech32 addresses.

Author: `They are also used to enable non-native segwit transactions using a process known as P2WPKH-in-P2SH.` Can't make sense of this. What does `non-native` mean, and  what are the benefits of this process?

Reviewer 2: Simplified.

For example: `3J98t1WpEZ73CNmQviecrnyiWrnqRhWNLy`

**SegWit address - [Bech32](https://en.bitcoin.it/wiki/Bech32){:target="_blank"}**<br/>
The *SegWit* address format (for segregated witness) was introduced in an update to the Bitcoin protocol that made changes to the transaction format. The technical name of this address format is Bech32, and is specified in [BIP173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki){:target="_blank"}. SegWit addresses are supported by the majority of software and hardware wallets, but a minority of exchanges. SegWit address adoption by applications and exchanges is tracked on [this page](https://en.bitcoin.it/wiki/Bech32_adoption). Any new bitcoin-application should support the use of SegWit addresses.

The benefits of SegWit addresses include more resistance to input errors (as they are case insensitive and use error-correcting codes), and lower transaction fees. The fee saving will depend on the type of transaction but for a common transfer of funds it can be in the 30-40% range.

> *SegWit addresses* start with `bc1` and are all lowercase and case insensitive. <br/>

Reviewer: ```suggestion
*SegWit addresses* start with `bc1` and is case insensitive. <br/>
```

Reviewer 2: Will fix typo (is/are) and commit separately.

> Bech32 is the *SegWit* address format (for segregated witness), and is specified in [BIP173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki){:target="_blank"}. They are supported by the majority of software and hardware wallets, but a minority of exchanges. While most exchanges enable sending funds to bech32 addresses, they don’t enable users to receive them with this format. Bech32 adoption by wallet-applications and exchanges is tracked on [this page](https://en.bitcoin.it/wiki/Bech32_adoption).

Author: `Bech32 is the *SegWit* address format`.  This always confuses me. Is it a Bech32 address or a SegWit address? Or both? Can we pick one?

Author: `they don’t enable users to receive them with this format`. Curious if there's a reason for this.

Reviewer 2: Simplified.

> The benefits of Bech32 addresses include more resistance to input errors (they are case insensitive, and use error-correcting codes), and lower transaction fees. Transaction fees are based on the number of bytes in your transaction and SegWit allows to move a significant part of the transaction to a new structure called witness. Bytes in the witness get a 75% discount. How much you save depends on your transaction. If it involves a lot of signatures that can be moved to the witness you save more.

Author: `benefits of Bech32 addresses`. This is only over Legacy addresses and not Script addresses, right?

Reviewer 2: I simplified this text but no, the benefit is over both. There is added confusion here however as you can 'nest' a segwit address in P2SH. https://blog.blockonomics.co/saving-transaction-fee-using-segwit-how-to-be-a-bitcoin-ninja-78d8416375db

- [BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki){:target="_blank"}: Mnemonic code for generating deterministic keys
- [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki){:target="_blank"}: Multi-account hierarchy for HD wallets
- [BIP49](https://github.com/bitcoin/bips/blob/master/bip-0049.mediawiki){:target="_blank"}: Derivation scheme for HD wallets using P2WPKH-nested-in-P2SH, nested SegWit
- [BIP84](https://github.com/bitcoin/bips/blob/master/bip-0084.mediawiki){:target="_blank"}: Derivation scheme for HD wallets using P2WPKH, native SegWit
- [BIP141](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki){:target="_blank"}: Segregated Witness, changes to transaction structure
- [BIP173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki){:target="_blank"}: *Bech32* standard for native segregated witness addresses 
- [BIP174](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki){:target="_blank"}: Partially Signed Bitcoin Transaction Format

> ## Change address

Author: Should this go under the Address" header further up, as another type of address?

Reviewer 2: Good idea, moved.

   alt-text = "Illustrations of different types of wallets."
   layout = "float-right-desktop"
   width = 400
   height = 300
%}

This term is often used interchangeably for very different things. A user can *download a bitcoin wallet* in the app store, and then the app may offer the option to “create a new wallet” on the home screen. A software wallet may allow for browsing and sending transactions, while a hardware wallet may be limited to key storage and signing functionality.

</div>

### Bitcoin wallet

All bitcoin-related data derived from and associated with a single recovery phrase. Most modern bitcoin wallets are HD wallets.

### HD wallet

Hierarchical Deterministic wallets, or HD wallets, can create infinite keypairs organized in a tree-structure (hierarchical) from a single (deterministic) controlling keypair. They were introduced and defined in [BIP32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki){:target="_blank"} and then expanded with [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki){:target="_blank"} which added the ability to handle multiple *accounts* in one wallet. 

> **Technicalities** - There are different bitcoin address standards; Legacy, Native SegWit (segregated witness) and Nested SegWit. They have to be held in different branches of the HD wallet but can be controlled by the same recovery phrase. Native SegWit uses Bech32 address formats specified in [BIP173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki){:target="_blank"}.

Reviewer: ```suggestion
**Technicalities** - There are different bitcoin address formats (see [address](#address)). They have to be held in different branches of the HD wallet but can be controlled by the same recovery phrase.
```

Author: `They have to be held in different branches`. This is tricky to understand. What's a branch? Does the user have to manually do this? How about `There are different bitcoin address formats (see [address](#address)), which can be generated from the same recovery phrase, using different methods.`?

Reviewer: > organized in a tree-structure

its tree branches, as stated in the paragraph above. i agree better analogies need to be better but "using different methods" brings more vagueness to the reader imo.

### Hardware wallet

A hardware device used to manage a bitcoin wallet.

### Wallet application

A software application used to manage a bitcoin wallet.

### Non-custodial / Custodial wallet

> A non-custodial wallet-product implies that the private key and/or recovery phrase is only known and exposed to the end-user. This means that transactions can never be made without the user’s permission. It also means that should the user forget or misplace their recovery phrase, the makers of the wallet-product are unable to help restore access to the funds in the bitcoin wallet. This guide is focused on non-custodial wallets.

Reviewer: ```suggestion
A non-custodial wallet-product implies that the private key and/or recovery phrase is in full control of the end-user. This means that transactions can never be made without the user's direct action. It also means that should the user forget or misplace their recovery phrase, wallet-product makers cannot help restore access to the funds in the bitcoin wallet. _This guide is focused on non-custodial wallets._
```

> With a custodial wallet-product, the users are not exposed to and in charge of securing the recovery phrase, often just requiring the users to sign in with their email and password. This means that the users have to trust the makers of the wallet-product to secure their recovery phrase and bitcoin. With a custodial wallet-product the makers of it are technically in control of their users funds. Most exchanges give users custodial wallets.

Reviewer: ```suggestion
With a custodial wallet-product, the users are not exposed to and in charge of securing the recovery phrase. Often custodial wallet-products require users to sign in with their email and password. Users have to trust the wallet-product makers to secure their recovery phrase and bitcoin. With a custodial wallet-product, the makers of it are technically in control of their users' funds. Most exchanges give users custodial wallets.
```

### Hot / Cold wallet

> *Hot* and *cold* are often used to describe the wallet in terms of being connected to the internet. Where a hot wallet is connected to the internet, a cold wallet is not. The idea is that a cold wallet is less susceptible to third-party theft over the internet. Most software wallet-products would be seen as hot (although some can be used just for signing on a device not connected to the internet) and most hardware wallet-products would be seen as cold (although they are sometimes connected for signing purposes). 

Reviewer: ```suggestion
*Hot* and *cold* describe a wallet in terms of being connected to the internet. Where a hot wallet is connected to the internet, a cold wallet is not. The idea is that a cold wallet is less susceptible to third-party theft over the internet. Most software wallet-products would be seen as hot (although some can be used just for signing on a device not connected to the internet) and most hardware wallet-products would be seen as cold (although they are sometimes connected for signing purposes). 
```

When the output of a transaction is used as the input of another transaction, it must be spent in its entirety. Sometimes the value of the output is higher than what the user wishes to pay. In this case, the bitcoin client generates a new Bitcoin address, and sends the difference back to this address. This is known as *change address*.

## Account

Just like wallet, the term account can also be used for very different things. In bitcoin wallets that follow the hierarchy described in [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki){:target="_blank"}, a bitcoin wallet can have multiple accounts, with each one having its own addresses. However, account is also oftentimes used for accounts with third-party service providers.

Differentiate between

- Bitcoin wallet account
- Service account

## Bitcoin / bitcoin

Bitcoin with a capital *B* is typically associated with Bitcoin the protocol and payment network. It is also often used to refer to as the ecosystem as a whole when writing about it in general terms. Bitcoin with a lowercase “b” written as “bitcoin” is usually associated specifically with bitcoin as the currency.

## Bitcoin Core

> A Bitcoin wallet often considered as a *reference implementation* of the Bitcoin protocol. It is the continuation of Satoshi Nakamoto's original code. More on the [Bitcoin Wiki](https://en.bitcoin.it/wiki/Bitcoin_Core).

Reviewer: ```suggestion
A Bitcoin wallet often considered as the *reference implementation* which also serves as the specification of the Bitcoin protocol. It is the continuation of Satoshi Nakamoto's original code. More on the [Bitcoin Wiki](https://en.bitcoin.it/wiki/Bitcoin_Core).
```

## BIP - Bitcoin improvement proposal

> A standardized design document format for suggesting improvements to Bitcoin. They are hosted on Github [here](https://github.com/bitcoin/bips){:target="_blank"}. Some important proposals to be aware of:

Reviewer: ```suggestion
A standardized technical document format for suggesting improvements to Bitcoin. They are hosted on Github [here](https://github.com/bitcoin/bips){:target="_blank"}. Some important proposals to be aware of:
```

- [BIP49](https://github.com/bitcoin/bips/blob/master/bip-0049.mediawiki){:target="_blank"}: Derivation scheme for HD wallets using P2WPKH-nested-in-P2SH, nested SegWit
- [BIP84](https://github.com/bitcoin/bips/blob/master/bip-0084.mediawiki){:target="_blank"}: Derivation scheme for HD wallets using P2WPKH, native SegWit
- [BIP141](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki){:target="_blank"}: Segregated Witness, changes to transaction structure
- [BIP173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki){:target="_blank"}: *Bech32* standard for native segregated witness addresses 
- [BIP174](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki){:target="_blank"}: Partially Signed Bitcoin Transaction Format

## Coin control

<div class="center" markdown="1">

{% include image.html
   image = "/assets/images/guide/glossary/coin-selection.jpg"
   retina = "/assets/images/guide/glossary/coin-selection@2x.jpg"
   alt-text = "Illustration of specific coins a user has received getting chosen for a new transacton."
   layout = "float-right-desktop"
   width = 400
   height = 400
%}

> The act of choosing which coins (really unspent outputs, or UTXOs) to forward to another address in a transaction. Wallet-applications can automatically choose the coins to use, but there are scenarios when users may want to manually choose what coins to send.

Reviewer: ```suggestion
The act of choosing which UTXO to forward to another address in a transaction. Wallet-applications can automatically choose the coins to use, but there are scenarios when users may want to manually choose what coins to send.
```

Reviewer 2: Can we use 'unspent transaction' or something more readable? 
UTXO needs its own explanation. (I did not write this myself though so not an expert here)

Reviewer 2: I will commit the suggestion and add 'unspent transaction output' and link to the UTXO explanation.

## Recovery phrase

_Also referred to as Seed, Mnemonic, and Backup phrase._

The controlling keypair of a bitcoin wallet can be derived from a *recovery phrase* of 12 words (or 18 or 24, which is less common) from a standardized list, defined in [BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki){:target="_blank"}. The recovery phrase provides full access to a bitcoin wallet as it contains the private key and is therefore very valuable. It’s extremely important to keep it safe, both from other people getting access to it and for yourself not to lose it by creating one or several backups of the phrase. In many applications most of this work falls on the user and it’s important to acknowledge the responsibility here of the makers of the application to ensure that the user is able and aware of how to securely store a recovery phrase backup.

Many wallet-applications work with HD Wallets and recovery phrases, and are interoperable, meaning you can change the application that can control your wallet should you wish (although there are some caveats depending on if they support just BIP32 or also BIP44). 

> **Technicalities** - Recovery of multisig-wallets needs both the extended public key and the recovery phrase of all paticipating keys as well as the master key fingerprint as defined by BIP32 concatenated with the derivation path of the public key. The derivation path is represented as 32-bit little endian unsigned integer indexes concatenated with each other. The number of 32 bit unsigned integer indexes must match the depth provided in the extended public key.

Reviewer 3: This is quite technical, wondering if we need to section at all?

Reviewer 2: Good point. I have a plan to add a page just about wallet recovery, so perhaps it's better suited there.

Author: Got a question about `Recovery of multisig-wallets needs both the extended public key and the recovery phrase of all paticipating keys as well as the master key fingerprint as defined by BIP32 concatenated with the derivation path of the public key.` Aren't the recovery phrases enough as those other values can be derived from them? And shouldn't it be avoided to bring the recovery phrases together in one place, as that would create a single point of failure?

TBH, I am struggling wrapping my head around multisig logic as I have not found simple explanations anywhere.

Reviewer: @danielnordh ive got some research for this that would be useful - lmk when you get started on it.

Reviewer 2: I cut this for now.

### HD wallet

Hierarchical Deterministic wallets, or HD wallets, can create infinite keypairs organized in a tree-structure (hierarchical) from a single (deterministic) controlling keypair. They were introduced and defined in [BIP32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki){:target="_blank"} and then expanded with [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki){:target="_blank"} which added the ability to handle multiple *accounts* in one wallet. 

**Technicalities** - There are different bitcoin address formats (see [address](#address)). They have to be held in different branches of the HD wallet but can be controlled by the same recovery phrase.

### Hardware wallet

A hardware device used to manage a bitcoin wallet. More on the [Hardware overview]({{ '/guide/getting-started/hardware/#hardware-wallets' | relative_url }}) page.

### Wallet application

A software application used to manage a bitcoin wallet. More on the [Hardware overview]({{ '/guide/getting-started/software/#wallets' | relative_url }}) page.

### Non-custodial / Custodial wallet

> _Also known as unhosted / self-Hosted / hosted wallet._

Reviewer 3: ```suggestion
_Also known as unhosted / self-hosted / hosted wallet._
```

# Glossary
{: .no_toc }

## Address

## Account

Just like wallet, the term account can also be used for very different things. In bitcoin wallets that follow the hierarchy described in BIP 44, a bitcoin wallet can have multiple accounts, with each one having its own addresses. However, account is also oftentimes used for accounts with third-party service providers.

Differentiate between

- Bitcoin wallet account
- Service account

## Bech32

## Bitcoin / bitcoin

## Bitcoin Core

> 

Reviewer 4: Bitcoin Core is an open source project which maintains and releases the primary Bitcoin (BTC) reference implementation software called "Bitcoin Core."

## Bitcoin Core GUI

> 

Reviewer 4: The graphical user interface (GUI) for the Bitcoin Core software which includes a full node and wallet functionality.
